File: tools/sptool/sptool.c
/*
 * Copyright (c) 2018-2020, Arm Limited. All rights reserved.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "sptool.h"

#define PAGE_SIZE		4096

/*
 * Entry describing Secure Partition package.
 */
struct sp_pkg_info {
	/* Location of the files in the host's RAM. */
	void *img_data, *pm_data;

	/* Size of the files. */
	uint32_t img_size, pm_size;

	/* Location of the binary files inside the package output file */
	uint32_t img_offset, pm_offset;
};

/*
 * List of input provided by user
 */
struct arg_list {
	char *usr_input;
	struct arg_list *next;
};

/* Align an address to a power-of-two boundary. */
static unsigned int align_to(unsigned int address, unsigned int boundary)
{
	unsigned int mask = boundary - 1U;

	if ((address & mask) != 0U)
		return (address + boundary) & ~mask;
	else
		return address;
}

/* Allocate a memory area of 'size' bytes and zero it. */
static void *xzalloc(size_t size, const char *msg)
{
	void *d;

	d = malloc(size);
	if (d == NULL) {
		fprintf(stderr, "error: malloc: %s\n", msg);
		exit(1);
	}

	memset(d, 0, size);

	return d;
}

/*
 * Write 'size' bytes from 'buf' into the specified file stream.
 * Exit the program on error.
 */
static void xfwrite(void *buf, size_t size, FILE *fp)
{
	if (fwrite(buf, 1, size, fp) != size) {
		fprintf(stderr, "error: Failed to write to output file.\n");
		exit(1);
	}
}

/*
 * Set the file position indicator for the specified file stream.
 * Exit the program on error.
 */
static void xfseek(FILE *fp, long offset, int whence)
{
	if (fseek(fp, offset, whence) != 0) {
		fprintf(stderr, "error: Failed to set file to offset 0x%lx (%d).\n",
		       offset, whence);
		perror(NULL);
		exit(1);
	}
}

/*
 * Free SP package structure
 */
static void cleanup(struct sp_pkg_info *sp)
{

	if (sp != NULL) {
		if (sp->img_data != NULL) {
			free(sp->img_data);
		}

		if (sp->pm_data != NULL) {
			free(sp->pm_data);
		}

		free(sp);

	}
}

/*
 * Free argument list structure
 */
static void freelist(struct arg_list *head)
{
	struct arg_list *tmp;

	while (head != NULL) {
		tmp = head;
		head = head->next;
		free(tmp);
	}
}

/*
 * Append user inputs in argument list structure
 */
static void append_user_input(struct arg_list **head, char *args)
{
	struct arg_list *tmp = *head;

	if (tmp == NULL) {
		tmp = xzalloc(sizeof(struct arg_list),
				"Failed to allocate arg_list struct");
		tmp->usr_input = args;
		*head = tmp;
	} else {
		while (tmp->next != NULL) {
			tmp = tmp->next;
		}
		tmp->next = xzalloc(sizeof(struct arg_list),
				"Failed to allocate arg_list struct");
		tmp = tmp->next;
		tmp->usr_input = args;
	}
}

/*
 * Allocate a buffer big enough to store the content of the specified file and
 * load the file into it. Fill 'size' with the file size. Exit the program on
 * error.
 */
static void load_file(const char *path, void **ptr, uint32_t *size)
{
	FILE *f = fopen(path, "rb");
	if (f == NULL) {
		fprintf(stderr, "error: %s couldn't be opened.\n", path);
		exit(1);
	}

	xfseek(f, 0, SEEK_END);
	*size = ftell(f);
	if (*size == 0) {
		fprintf(stderr, "error: Size of %s is 0\n", path);
		exit(1);
	}

	rewind(f);

	*ptr = malloc(*size);
	if (*ptr == NULL) {
		fprintf(stderr, "error: Not enough memory to load %s\n", path);
		exit(1);
	}

	if (fread(*ptr, *size, 1, f) != 1) {
		fprintf(stderr, "error: Couldn't read %s\n", path);
		exit(1);
	}

	fclose(f);
}

/*
 * Parse the string containing input payloads and fill in the
 * SP Package data structure.
 */
static void load_sp_pm(char *path, struct sp_pkg_info **sp_out)
{
	struct sp_pkg_info *sp_pkg;

	char *split_mark = strstr(path, ":");

	*split_mark = '\0';

	char *sp_path = path;
	char *pm_path = split_mark + 1;

	sp_pkg = xzalloc(sizeof(struct sp_pkg_info),
		"Failed to allocate sp_pkg_info struct");

	load_file(pm_path, &sp_pkg->pm_data, &sp_pkg->pm_size);
	printf("\nLoaded SP Manifest file %s (%u bytes)\n", pm_path, sp_pkg->pm_size);

	load_file(sp_path, &sp_pkg->img_data, &sp_pkg->img_size);
	printf("Loaded SP Image file %s (%u bytes)\n", sp_path, sp_pkg->img_size);

	*sp_out = sp_pkg;
}

/*
 * Write SP package data structure into output file.
 */
static void output_write(const char *path, struct sp_pkg_info *sp, bool header)
{
	struct sp_pkg_header sp_header_info;
	unsigned int file_ptr = 0;

	FILE *f = fopen(path, "wb");
	if (f == NULL) {
		fprintf(stderr, "error: Failed to open %s\n", path);
		exit(1);
	}

	/* Reserve Header size */
	if (header) {
		file_ptr = sizeof(struct sp_pkg_header);
	}

	/* Save partition manifest */
	xfseek(f, file_ptr, SEEK_SET);
	printf("Writing SP Manifest at offset 0x%x (%u bytes)\n",
	       file_ptr, sp->pm_size);

	sp->pm_offset = file_ptr;
	xfwrite(sp->pm_data, sp->pm_size, f);

	/* Save partition image aligned to Page size */
	file_ptr = align_to((sp->pm_offset + sp->pm_size), PAGE_SIZE);
	xfseek(f, file_ptr, SEEK_SET);
	printf("Writing SP Image at offset 0x%x (%u bytes)\n",
	       file_ptr, sp->img_size);

	sp->img_offset = file_ptr;
	xfwrite(sp->img_data, sp->img_size, f);

	/* Finally, write header, if needed */
	if (header) {
		sp_header_info.magic = SECURE_PARTITION_MAGIC;
		sp_header_info.version = 0x1;
		sp_header_info.img_offset = sp->img_offset;
		sp_header_info.img_size = sp->img_size;
		sp_header_info.pm_offset = sp->pm_offset;
		sp_header_info.pm_size = sp->pm_size;

		xfseek(f, 0, SEEK_SET);

		printf("Writing package header\n");

		xfwrite(&sp_header_info, sizeof(struct sp_pkg_header), f);
	}

	/* All information has been written now */
	printf("\nsptool: Built Secure Partition blob %s\n", path);

	fclose(f);
}

static void usage(void)
{
	printf("usage: sptool ");
#ifdef VERSION
	printf(VERSION);
#else
	/* If built from sptool directory, VERSION is not set. */
	printf("version unknown");
#endif
	printf(" [<args>]\n\n");

	printf("This tool takes as input set of image binary files and the\n"
	       "partition manifest blobs as input and generates set of\n"
	       "output package files\n"
	       "Usage example: sptool -i sp1.bin:sp1.dtb -o sp1.pkg\n"
	       "                      -i sp2.bin:sp2.dtb -o sp2.pkg ...\n\n");
	printf("Commands supported:\n");
	printf("  -o <path>            Set output file path.\n");
	printf("  -i <sp_path:pm_path> Add Secure Partition image and\n"
	       "                       Manifest blob (specified in two paths\n"
	       "                       separated by a colon).\n");
	printf("  -n                   Generate package without header\n");
	printf("  -h                   Show this message.\n");
	exit(1);
}

int main(int argc, char *argv[])
{
	struct sp_pkg_info *sp_pkg = NULL;
	struct arg_list *in_head = NULL;
	struct arg_list *out_head = NULL;
	struct arg_list *in_list = NULL;
	struct arg_list *out_list = NULL;
	unsigned int match_counter = 0;
	bool need_header = true;

	int ch;

	if (argc <= 1) {
		fprintf(stderr, "error: File paths must be provided.\n\n");
		usage();
		return 1;
	}

	while ((ch = getopt(argc, argv, "hni:o:")) != -1) {
		switch (ch) {
		case 'i':
			append_user_input(&in_head, optarg);
			match_counter++;
			break;
		case 'o':
			append_user_input(&out_head, optarg);
			match_counter--;
			break;
		case 'n':
			need_header = false;
			break;
		case 'h':
		default:
			usage();
		}
	}

	if (match_counter) {
		fprintf(stderr, "error: Input/Output count mismatch.\n\n");
		freelist(in_head);
		freelist(out_head);
		usage();
		return 1;
	}

	in_list = in_head;
	out_list = out_head;
	while (in_list != NULL) {
		load_sp_pm(in_list->usr_input, &sp_pkg);
		output_write(out_list->usr_input, sp_pkg, need_header);
		in_list = in_list->next;
		out_list = out_list->next;
	}

	argc -= optind;
	argv += optind;

	cleanup(sp_pkg);
	freelist(in_head);
	freelist(out_head);

	return 0;
}

File: tools/sptool/sptool.py
#!/usr/bin/python3
# Copyright (c) 2022, Arm Limited. All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause

#
# Copyright 2022 The Hafnium Authors.
#
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/BSD-3-Clause.

"""
Script which generates a Secure Partition package.
https://trustedfirmware-a.readthedocs.io/en/latest/components/secure-partition-manager.html#secure-partition-packages
"""

import argparse
from collections import namedtuple
import sys
from shutil import copyfileobj
import os

HF_PAGE_SIZE = 0x1000 # bytes
HEADER_ELEMENT_BYTES = 4 # bytes
MANIFEST_IMAGE_SPLITTER=':'
PM_OFFSET_DEFAULT = "0x1000"
IMG_OFFSET_DEFAULT = "0x4000"

def split_dtb_bin(i : str):
    return i.split(MANIFEST_IMAGE_SPLITTER)

def align_to_page(n):
    return HF_PAGE_SIZE * \
          (round(n / HF_PAGE_SIZE) + \
           (1 if n % HF_PAGE_SIZE else 0))

def to_bytes(value):
    return int(value).to_bytes(HEADER_ELEMENT_BYTES, 'little')

class SpPkg:
    def __init__(self, pm_path : str, img_path : str, pm_offset: int,
                 img_offset: int):
        if not os.path.isfile(pm_path) or not os.path.isfile(img_path):
            raise Exception(f"Parameters should be path.  \
                              manifest: {pm_path}; img: {img_path}")
        self.pm_path = pm_path
        self.img_path = img_path
        self._SpPkgHeader = namedtuple("SpPkgHeader",
                             ("magic", "version",
                              "pm_offset", "pm_size",
                              "img_offset", "img_size"))

        if pm_offset >= img_offset:
            raise ValueError("pm_offset must be smaller than img_offset")

        is_hfpage_aligned = lambda val : val % HF_PAGE_SIZE == 0
        if not is_hfpage_aligned(pm_offset) or not is_hfpage_aligned(img_offset):
           raise ValueError(f"Offsets provided need to be page aligned: pm-{pm_offset}, img-{img_offset}")

        if img_offset - pm_offset < self.pm_size:
            raise ValueError(f"pm_offset and img_offset do not fit the specified file:{pm_path})")

        self.pm_offset = pm_offset
        self.img_offset = img_offset

    def __str__(self):
        return \
        f'''--SP package Info--
        header:{self.header}
        pm: {self.pm_path}
        img: {self.img_path}
        '''

    @property
    def magic(self):
        return "SPKG".encode()

    @property
    def version(self):
        return 0x2

    @property
    def pm_size(self):
        return os.path.getsize(self.pm_path)

    @property
    def img_size(self):
        return os.path.getsize(self.img_path)

    @property
    def header(self):
        return self._SpPkgHeader(
                self.magic,
                self.version,
                self.pm_offset,
                self.pm_size,
                self.img_offset,
                self.img_size)

    @property
    def header_size(self):
        return len(self._SpPkgHeader._fields)

    def generate(self, f_out : str):
        with open(f_out, "wb+") as output:
            for h in self.header:
                to_write = h if type(h) is bytes else to_bytes(h)
                output.write(to_write)
            output.seek(self.pm_offset)
            with open(self.pm_path, "rb") as pm:
                copyfileobj(pm, output)
            output.seek(self.img_offset)
            with open(self.img_path, "rb") as img:
                copyfileobj(img, output)

def Main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", required=True,
                        help="path to partition's image and manifest separated by a colon.")
    parser.add_argument("--pm-offset", required=False, default=PM_OFFSET_DEFAULT,
                        help="set partitition manifest offset.")
    parser.add_argument("--img-offset", required=False, default=IMG_OFFSET_DEFAULT,
                        help="set partition image offset.")
    parser.add_argument("-o", required=True, help="set output file path.")
    parser.add_argument("-v", required=False, action="store_true",
                        help="print package information.")
    args = parser.parse_args()

    if not os.path.exists(os.path.dirname(args.o)):
        raise Exception("Provide a valid output file path!\n")

    image_path, manifest_path = split_dtb_bin(args.i)
    pm_offset = int(args.pm_offset, 0)
    img_offset = int(args.img_offset, 0)
    pkg = SpPkg(manifest_path, image_path, pm_offset, img_offset)
    pkg.generate(args.o)

    if args.v is True:
        print(pkg)

    return 0

if __name__ == "__main__":
    sys.exit(Main())
